### ISSUE TITLE
Deprecate some compiler directives

### ISSUE THREAD

**Issue by ark120202** [TypeScriptToLua/TypeScriptToLua#652]
* `@compileMembersOnly` - too generic, targeted transforms cover it better and avoid name duplication (now it's `EnumName.ENUM_NAME_ENUM_MEMBER`).
* `@extension` - too generic, originally it was designed as a workaround to add members to existing classes, but interface merging + simple property set works, and it's exactly like it's done in JS.
* `@pureAbstract` - it was added only for dota, interface merging covers it.
---
* `@metaExtension` - looks like it was added only for a single environment, I think we might reevaluate the need for it.
* `@customConstructor` - I personally don't see much point in it, wouldn't it be better to type it as an interface and function?
* `@phantom` - what was an original motivation for it? It looks similar to #65 and #645.

**Comment by lolleko** [comment]
I am in favour of deprecating `@extension` since its broken (in part #492) for quite a while now and I don't think fixing/maintaining it is worth it. Since fixing it would require 2 different class translation modes (with/without `@extension`)

Same for `@pureAbstract` since we now worked out viable alternatives I don't think its worth maintaining it anymore.
`@metaExtension` can also be achieved differently, similar to pureAbstract.

As for `@compileMembersOnly` and `@customConstructor` while they are not required, I think they are very useful to improve code quality. In addition these are widely used in projects, so deprecating them would require a lot of people to change the code (I don't think that is the case for most of the other annotations.)

`@phantom`: Although I can't recall what the specific use case for phantom was, being able to create `do end` scopes in Lua sounds useful. In addition `@phantom` is really cheap to maintain.

**TL;DR:**

Deprecate: `@extension, @pureAbtract, @metaExtension`
Keep (at least for now): `@compileMembersOnly, @customConstructor, phantom`

---

For the deprecation process I would start by emitting a warnings In the next version: `Warning XYZ annotation is deprecated and will be removed in a future update`. In couple of months I would remove all of the code related to those annotations and change the warning to an error: `Error XYZ annotation is no longer supported`.
